opam accepts switch names that create unusable switches

Hi,

Opam can create switches with some characters in their names that create unusable switches:

### colons

When a switch name contains a colon, `PATH` is extended with this switch's bin directory, but since this is a colon-separated variable (with no escape mechanism), it fails to find the right binaries:

```
% opam switch create test:a 4.09.1 
...
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed ocaml-base-compiler.4.09.1
∗ installed ocaml-config.1
[ERROR] The compilation of ocaml failed at
        "/home/etienne/.opam/opam-init/hooks/sandbox.sh build ocaml
        /home/etienne/.opam/test:a/share/ocaml-config/gen_ocaml_config.ml 4.09.1
        ocaml".

#=== ERROR while compiling ocaml.4.09.1 =======================================#
# context     2.0.0 | linux/x86_64 | ocaml-base-compiler.4.09.1 | https://opam.ocaml.org/#9e7e7e45
# path        ~/.opam/test:a/.opam-switch/build/ocaml.4.09.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build ocaml /home/etienne/.opam/test
:a/share/ocaml-config/gen_ocaml_config.ml 4.09.1 ocaml
# exit-code   1
# env-file    ~/.opam/log/ocaml-104246-ad6566.env
# output-file ~/.opam/log/ocaml-104246-ad6566.out
### output ###
# OCaml version mismatch: 4.08.1, expected 4.09.1
```

The reason for the error message is that it tries to find `ocaml` in `PATH`, but because there's a colon in there it cannot find it and falls back to my system compiler.

This happens in a local switch as well if the directory contains colons at any level.

### shell metacharacters (backticks, dollar sign)

When a switch name (or in the case of a local switch, a directory name) contains characters that are interpreted by the shell such as backticks or dollar signs, they are interpreted during the installation. For example, a switch created in a directory named `id` (backtick, i, d, backtick) will try to create a switch named by the output of this command, and a switch named `xmessage hello` (backtick, x, ..., o, backtick) will open popup window during switch creation. A similar behavior can be observed with dollar signs. This suggests that somewhere a command is built with double quotes and passed to the shell.

### Possible security concerns

Opam is primarily a development tool, so it's expected that it will run external code or install binaries. There might be scenarios where an attacker can control a directory name or a global switch name but I don't think that should be an important concern.

Thanks!

--- a/master_changes.md
+++ b/master_changes.md
@@ -23,6 +23,7 @@ Possibly scripts breaking changes are prefixed with ✘
 ## Switch
   * Fix Not_found with `opam switch create . --deps` [#4151 @AltGr]
   * Package Var: resolve self `name` variable for orphan packages [#4228 @rjbou - fix #4224]
+  * ✘ Reject (shell) character on switch names [#4237 @rjbou - fix #4231]
 
 ## Pin
   * Don't keep unpinned package version if it exists in repo [#4073 @rjbou - fix #3630]
@@ -44,16 +45,23 @@ Possibly scripts breaking changes are prefixed with ✘
   * Fix install command dryrun [#4200 @rjbou]
 
 
+## Env
+  * Fix `OPAMSWITCH` empty string setting, consider as unset [#4237 @rjbou]
+
 ## Remove
   * Fix autoremove env var handling [#4219 @rjbou - fix #4217]
 
 ## Repository management
   * Fix temp files repository cleaning [#4197 @rjbou]
 
+## Opam installer
+  * For paths, remove use of empty switch in favor of a context-less module [#4237 @rjbou]
+
 ## Internal
   * Disable chrono when timestamps are disables [#4206 @rjbou]
   * Expose some functionality in the `OpamAction`, `OpamPath` and `OpamSwitchState`
-    modules for use without a `switch` value [#4147]
+    modules for use without a `switch` value [#4147 @timberston]
+    *Path: introduce a functor to permit replicating switch layout in different contexts
 
 ## Test
   * Add show cram test [#4206 @rjbou]

--- a/src/client/opamConfigCommand.ml
+++ b/src/client/opamConfigCommand.ml
@@ -182,7 +182,7 @@ let env gt switch ?(set_opamroot=false) ?(set_opamswitch=false)
         (OpamFile.Config.switch gt.config)
     in
     match OpamStd.Config.env_string "SWITCH" with
-    | None ->
+    | None | Some "" ->
       Some (OpamStateConfig.resolve_local_switch gt.root switch) <> default
     | Some s ->
       OpamStateConfig.resolve_local_switch gt.root (OpamSwitch.of_string s) <>

--- a/src/format/opamSwitch.ml
+++ b/src/format/opamSwitch.ml
@@ -20,17 +20,18 @@ let is_external s =
 let external_dirname = "_opam"
 
 let check s =
-  let re =
-    Re.(compile @@
-        seq [
-          bol;
-          opt @@ seq [ wordc ; char ':'; set "/\\" ];
-          rep @@ diff any @@ set "<>!`$():";
-          eol
-        ])
-  in
-  (try ignore @@ Re.exec re s with Not_found ->
-     failwith (Printf.sprintf "Invalid character in switch name %S" s));
+  if String.compare s "" = 0 &&
+    let re =
+      Re.(compile @@
+          seq [
+            bol;
+            opt @@ seq [ wordc ; char ':'; set "/\\" ];
+            rep @@ diff any @@ set "<>!`$():";
+            eol
+          ])
+    in
+    (try ignore @@ Re.exec re s; true with Not_found -> false) then
+    failwith (Printf.sprintf "Invalid character in switch name %S" s);
   s
 
 let of_string s =

--- a/src/format/opamSwitch.ml
+++ b/src/format/opamSwitch.ml
@@ -19,12 +19,28 @@ let is_external s =
 
 let external_dirname = "_opam"
 
+let check s =
+  let re =
+    Re.(compile @@
+        seq [
+          bol;
+          opt @@ seq [ wordc ; char ':'; set "/\\" ];
+          rep @@ diff any @@ set "<>!`$():";
+          eol
+        ])
+  in
+  (try ignore @@ Re.exec re s with Not_found ->
+     failwith (Printf.sprintf "Invalid character in switch name %S" s));
+  s
+
 let of_string s =
+  check @@
   if is_external s then OpamFilename.Dir.(to_string (of_string s))
   else s
 
 let of_dirname d =
   let s = OpamFilename.Dir.to_string d in
+  check @@
   try
     let swdir = Unix.readlink (Filename.concat s external_dirname) in
     let swdir =

--- a/src/tools/opam_installer.ml
+++ b/src/tools/opam_installer.ml
@@ -156,7 +156,16 @@ let script_commands project_root ochan =
    where [dest src dst] returns the destination of a file with a
   ["src" {"dst"}] line in the .install *)
 let iter_install f instfile o =
-  let module D = OpamPath.Switch.Default in
+  let open OpamFilename.Op in
+  let module D =
+  struct
+    include OpamPath.Switch.DefaultF (struct
+        type ctx = unit
+        let root d _ = d
+        let lib_dir t a = root t a / "lib"
+      end)
+  end
+  in
   let module S = OpamFile.Dot_install in
   let dest ?fix dir =
     let dir = OpamStd.Option.default dir fix in
@@ -164,18 +173,14 @@ let iter_install f instfile o =
       OpamFilename.create dir
         (OpamStd.Option.default (OpamFilename.basename src) dst)
   in
-  let dest_global ?fix instdir_f =
-    dest ?fix (instdir_f o.prefix (OpamSwitch.of_string ""))
-  in
+  let dest_global ?fix instdir_f = dest ?fix (instdir_f o.prefix ()) in
   let dest_pkg ?fix instdir_f =
     let fix =
-      OpamStd.Option.map
-        OpamFilename.Op.(fun d ->
-            d / OpamPackage.Name.to_string o.pkgname)
+      OpamStd.Option.map (fun d ->
+          d / OpamPackage.Name.to_string o.pkgname)
         fix
     in
-    dest ?fix
-      (instdir_f o.prefix (OpamSwitch.of_string "") o.pkgname)
+    dest ?fix (instdir_f o.prefix () o.pkgname)
   in
   List.iter f
     [ dest_global                 D.bin,      S.bin instfile,        true;

